### PR TITLE
[Rebase & FF] 202405: PCD Loader feature cherry-pick

### PR DESCRIPTION
**Pull request by cfernald**
## Description

Cherry-pick of change for PCD loader feature.

6057adbac4

- [x] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

N/A

## Integration Instructions

N/A

### CODE REVIEW

File: MdeModulePkg/Include/Library/PcdDatabaseLoaderLib.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/** @file
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+  Library interface to load the PCD Database from the Platform OEM Data FV.
+**/
+
+#ifndef _PCD_DATABASE_LOADER_LIB_H_
```
Comment (os-d) [inline MdeModulePkg/Include/Library/PcdDatabaseLoaderLib.h]:
```suggestion
#ifndef PCD_DATABASE_LOADER_LIB_H_
```
nit

File: MdeModulePkg/Library/PcdDatabaseLoaderLib/Dxe/PcdDatabaseLoaderLibDxe.c [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/** @file -- PcdDatabaseLoaderLibDxe.c
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline MdeModulePkg/Library/PcdDatabaseLoaderLib/Dxe/PcdDatabaseLoaderLibDxe.c]:
```suggestion
  Copyright (c) Microsoft Corporation.
```
Please check in all the files.

File: MdeModulePkg/Universal/PCD/Dxe/Service.c [hunk 3]
```diff
@@ -738,19 +738,10 @@ LocateExPcdBinary (
...
-             EFI_SECTION_RAW,
-             0,
-             (VOID **)&mDxePcdDbBinary,
-             &mDxePcdDbSize
-             );
-  ASSERT_EFI_ERROR (Status);
+  // MU_CHANGE begin
+  mDxePcdDbBinary = (DXE_PCD_DATABASE *)PcdDatabaseLoaderLoad (NULL);
+  ASSERT (mDxePcdDbBinary != NULL);
+  // MU_CHANGE end
```
Comment (os-d) [inline MdeModulePkg/Universal/PCD/Dxe/Service.c:757]: Should we change this to comment out the original code, the current MU_CHANGE style?

File: MdeModulePkg/Universal/PCD/Pei/Pcd.c [hunk 4]
```diff
@@ -309,8 +309,8 @@ EndOfPeiSignalPpiNotifyCallback (
   //
   // Find PEI PcdDb and Build second PcdDB GuidHob
   //
-  Status = PeiServicesFfsFindSectionData (EFI_SECTION_RAW, FileHandle, &PcdDb);
-  ASSERT_EFI_ERROR (Status);
+  PcdDb = (PEI_PCD_DATABASE *)PcdDatabaseLoaderLoad (FileHandle);     // MU_CHANGE
+  ASSERT (PcdDb != NULL);                                             // MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/Universal/PCD/Pei/Pcd.c]: Same with this one

File: MdeModulePkg/Universal/PCD/Pei/Pcd.c [hunk 5]
```diff
@@ -614,8 +614,8 @@ PeiPcdSetSku (
     //
     // Find the delta data between the different Skus
     //
-    Status = PeiServicesFfsFindSectionData (EFI_SECTION_RAW, FileHandle, &PcdDb);
-    ASSERT_EFI_ERROR (Status);
+    PcdDb = (PEI_PCD_DATABASE *)PcdDatabaseLoaderLoad (FileHandle);   // MU_CHANGE
+    ASSERT (PcdDb != NULL);                                           // MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/Universal/PCD/Pei/Pcd.c]: And here

File: MdeModulePkg/Universal/PCD/Pei/Service.c [hunk 6]
```diff
@@ -389,15 +389,14 @@ LocateExPcdBinary (
...
-  PcdDb = NULL;
+  // MU_CHANGE begin
+  VOID  *PcdDb;
 
   ASSERT (FileHandle != NULL);
 
-  Status = PeiServicesFfsFindSectionData (EFI_SECTION_RAW, FileHandle, &PcdDb);
-  ASSERT_EFI_ERROR (Status);
+  PcdDb = (PEI_PCD_DATABASE *)PcdDatabaseLoaderLoad (FileHandle);
+  ASSERT (PcdDb != NULL);
```
Comment (os-d) [inline MdeModulePkg/Universal/PCD/Pei/Service.c:404]: And for this file

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/Library/PcdDatabaseLoaderLib/Dxe/PcdDatabaseLoaderLibDxe.inf [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+#
+# The following information is for reference only and not required by the build tools.
+#
+# VALID_ARCHITECTURES = IA32 X64 AARCH64
+#
+
+[LibraryClasses]
+  DebugLib
+  DxeServicesLib
+  PcdDatabaseLoaderLib
```
Comment (makubacki) [inline MdeModulePkg/Library/PcdDatabaseLoaderLib/Dxe/PcdDatabaseLoaderLibDxe.inf]: It seems weird this lists the library class it's implementing as a dependency.

```suggestion
```

File: MdeModulePkg/Universal/PCD/Dxe/Service.c (hunk 3 above)
Comment (makubacki) [reply MdeModulePkg/Universal/PCD/Dxe/Service.c:757]: I think so.

**Review by makubacki — APPROVED** [review]

**Pull request merged**